File: filters/vacancy_filter.py
"""
Модуль фильтрации вакансий.

Применяет набор фильтров (зарплата, стоп-слова, тип занятости,
ключевые слова) к списку вакансий перед публикацией в канал.
Все параметры фильтрации берутся из config.FILTERS.
"""

import sys
import os

sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..'))

from config import FILTERS
from loguru import logger


def is_valid_vacancy(vacancy: dict) -> bool:
    """Проверяет, проходит ли вакансия все фильтры.

    Вакансия — нормализованный словарь с полями:
        title, company, salary_from, salary_to, schedule,
        description, url, source, vacancy_id.

    Args:
        vacancy: словарь с данными вакансии.

    Returns:
        True — вакансия прошла все фильтры, False — отсеяна.
    """
    title = vacancy.get("title", "")

    # ── 1. Проверка зарплаты ──────────────────────────────────
    salary_from = vacancy.get("salary_from")
    if salary_from is not None and salary_from < FILTERS["salary_min"]:
        logger.debug(
            "Вакансия отсеяна по зарплате: «{}» — {} < {}",
            title,
            salary_from,
            FILTERS["salary_min"],
        )
        return False

    # ── 2. Проверка стоп-слов ─────────────────────────────────
    text_combined = f"{title} {vacancy.get('description', '')}".lower()

    for stop_word in FILTERS["keywords_exclude"]:
        if stop_word.lower() in text_combined:
            logger.debug(
                "Вакансия отсеяна по стоп-слову «{}»: «{}»",
                stop_word,
                title,
            )
            return False

    # ── 3. Проверка типа занятости (schedule) ─────────────────
    schedule_value = vacancy.get("schedule", "")
    # Поддерживаем оба ключа из config: "schedule" и "schedule_keywords"
    schedule_keywords = FILTERS.get("schedule", FILTERS.get("schedule_keywords", []))

    if schedule_value:
        schedule_lower = schedule_value.lower()
        if not any(kw.lower() in schedule_lower for kw in schedule_keywords):
            logger.debug(
                "Вакансия отсеяна по типу занятости: «{}» — schedule='{}'",
                title,
                schedule_value,
            )
            return False

    # ── 4. Проверка включающих ключевых слов ──────────────────
    keywords_include = FILTERS.get("keywords_include", [])
    if keywords_include:
        if not any(kw.lower() in text_combined for kw in keywords_include):
            logger.debug(
                "Вакансия отсеяна — не найдены ключевые слова: «{}»",
                title,
            )
            return False

    return True


def filter_vacancies(vacancies: list[dict]) -> list[dict]:
    """Фильтрует список вакансий, оставляя только прошедшие все проверки.

    Args:
        vacancies: исходный список словарей-вакансий.

    Returns:
        Отфильтрованный список вакансий.
    """
    count_before = len(vacancies)

    try:
        result = [v for v in vacancies if is_valid_vacancy(v)]
    except Exception as exc:
        logger.error("Ошибка при фильтрации вакансий: {}", exc)
        return []

    count_after = len(result)
    logger.info(
        "Фильтрация завершена: {} → {} (отсеяно {})",
        count_before,
        count_after,
        count_before - count_after,
    )
    return result

File: src/filters/vacancyFilter.ts
import { config } from '../config/config.js';
import { logger } from '../utils/logger.js';
import { Vacancy } from '../db/database.js';

export function isValidVacancy(vacancy: Vacancy): boolean {
  const title = String(vacancy.title || '').toLowerCase();
  const description = String(vacancy.description || '').toLowerCase();
  const requirements = String(vacancy.requirements || '').toLowerCase();
  const schedule = String(vacancy.schedule || '').toLowerCase();

  const combinedText = `${title} ${description} ${requirements}`;

  // 1. Check salary (salary_from)
  if (vacancy.salary_from && vacancy.salary_from < config.FILTERS.salary_min) {
    logger.debug(
      `Filtered out (Salary too low: ${vacancy.salary_from} < ${config.FILTERS.salary_min}): "${vacancy.title}"`
    );
    return false;
  }

  // 2. Check excluded keywords (stop words)
  for (const stopWord of config.FILTERS.keywords_exclude) {
    if (combinedText.includes(stopWord)) {
      logger.debug(
        `Filtered out (Stop word matched "${stopWord}"): "${vacancy.title}"`
      );
      return false;
    }
  }

  // 3. Check schedule type (remote-only check)
  if (schedule) {
    const hasRemoteKeyword = config.FILTERS.schedule.some((keyword) =>
      schedule.includes(keyword)
    );
    if (!hasRemoteKeyword) {
      logger.debug(
        `Filtered out (Not remote schedule: "${vacancy.schedule}"): "${vacancy.title}"`
      );
      return false;
    }
  } else {
    // If no schedule information is provided, check if the title or description mentions remote work
    const textHasRemoteKeyword = config.FILTERS.schedule.some(
      (keyword) => title.includes(keyword) || description.includes(keyword)
    );
    if (!textHasRemoteKeyword) {
      logger.debug(
        `Filtered out (No remote work keywords found in text): "${vacancy.title}"`
      );
      return false;
    }
  }

  // 4. Check included keywords (if defined)
  if (config.FILTERS.keywords_include.length > 0) {
    const hasIncludeKeyword = config.FILTERS.keywords_include.some((keyword) =>
      combinedText.includes(keyword.toLowerCase())
    );
    if (!hasIncludeKeyword) {
      logger.debug(
        `Filtered out (Missing required keywords): "${vacancy.title}"`
      );
      return false;
    }
  }

  return true;
}

export function filterVacancies(vacancies: Vacancy[]): Vacancy[] {
  const countBefore = vacancies.length;
  const filtered = vacancies.filter(isValidVacancy);
  const countAfter = filtered.length;

  logger.info(`Filtering summary: ${countBefore} -> ${countAfter} (filtered out ${countBefore - countAfter})`);
  return filtered;
}
